File: library/src/main/java/com/alamkanak/weekview/WeekViewViewState.java
package com.alamkanak.weekview;

import java.util.Calendar;

import static com.alamkanak.weekview.Constants.HOURS_PER_DAY;
import static java.lang.Math.max;

final class WeekViewViewState {

    Calendar scrollToDay = null;
    int scrollToHour = -1;

    boolean isFirstDraw = true;
    boolean areDimensionsInvalid = true;

    Calendar firstVisibleDay;
    Calendar lastVisibleDay;

    boolean shouldRefreshEvents;

    void update(WeekViewConfig config, UpdateListener listener) {
        if (!areDimensionsInvalid) {
            return;
        }

        final float totalHeaderHeight = config.drawingConfig.headerHeight
                + config.headerRowPadding * 2
                + config.drawingConfig.headerMarginBottom;

        final int height = WeekView.getViewHeight();

        config.effectiveMinHourHeight = max(
                config.minHourHeight,
                (int) ((height - totalHeaderHeight) / HOURS_PER_DAY)
        );

        areDimensionsInvalid = false;
        if (scrollToDay != null) {
            listener.goToDate(scrollToDay);
        }

        areDimensionsInvalid = false;
        if (scrollToHour >= 0) {
            listener.goToHour(scrollToHour);
        }

        scrollToDay = null;
        scrollToHour = -1;
        areDimensionsInvalid = false;
    }

    void invalidate() {
        areDimensionsInvalid = false;
    }

    interface UpdateListener {
        void goToDate(Calendar date);
        void goToHour(int hour);
    }

}

File: library/src/main/java/com/alamkanak/weekview/WeekViewViewState.kt
package com.alamkanak.weekview

import java.lang.Math.max
import java.util.*

internal class WeekViewViewState(
    private val configWrapper: WeekViewConfigWrapper,
    private val listener: Listener
) {

    var scrollToDay: Calendar? = null
    var scrollToHour: Int? = null

    var isFirstDraw = true
    var areDimensionsInvalid = true

    var firstVisibleDay: Calendar? = null
    var lastVisibleDay: Calendar? = null

    var shouldRefreshEvents: Boolean = false
    var requiresPostInvalidateOnAnimation: Boolean = false

    fun update() {
        val totalHeaderHeight = configWrapper.getTotalHeaderHeight()

        val totalHeight = WeekView.height
        val dynamicHourHeight = ((totalHeight - totalHeaderHeight) / configWrapper.hoursPerDay).toInt()

        configWrapper.effectiveMinHourHeight = max(configWrapper.minHourHeight, dynamicHourHeight)

        areDimensionsInvalid = false
        scrollToDay?.let {
            isFirstDraw = false
            listener.goToDate(it)
        }

        areDimensionsInvalid = false
        scrollToHour?.let {
            listener.goToHour(it)
        }

        scrollToDay = null
        scrollToHour = null
        areDimensionsInvalid = false
    }

    fun invalidate() {
        areDimensionsInvalid = false
    }

    internal interface Listener {
        fun goToDate(date: Calendar)
        fun goToHour(hour: Int)
    }

}
